modules/rosetta/src/main/scala/org/tessellation/rosetta/domain/construction/ConstructionService.scala

```diff
@@ -4,129 +4,192 @@ import cats.data.{EitherT, NonEmptyList, NonEmptySet}
+        } yield ConstructionParse.ParseResult(operations, none)
+
+      def parseTransaction(hex: Hex, isSigned: Boolean): EitherT[F, ConstructionError, ConstructionParse.ParseResult] =
+        if (isSigned) {
+          parseSignedTransaction(hex)
+        } else {
+          parseUnsignedTransaction(hex)
         }
 
-  }
+      def combineTransaction(hex: Hex, signature: RosettaSignature): EitherT[F, ConstructionError, Hex] =
+        EitherT
+          .fromEither(KryoSerializer[F].deserialize[DAGTransaction](hex.toBytes))
+          .leftMap(_ => MalformedTransaction)
+          .flatMap { transaction =>
+            EitherT {
+              signature.publicKey.hexBytes.toPublicKeyByEC
+                .map(pk => SignatureProof(pk.toId, Signature(signature.hexBytes)).asRight[ConstructionError])
+            }.flatMap { proof =>
+              EitherT
+                .fromEither(
+                  KryoSerializer[F]
+                    .serialize(Signed[DAGTransaction](transaction, NonEmptySet.of(proof)))
+                )
+                .leftMap(_ => MalformedTransaction.asInstanceOf[ConstructionError])
+                .map(Hex.fromBytes(_))
+            }
+          }
+
+      override def getPayloads(
+        operations: NonEmptyList[Operation],
+        metadataResult: MetadataResult
+      ): EitherT[F, ConstructionError, PayloadsResult] =
+        EitherT(
+          saltGen().map(transactionSalt =>
+            for {
+              (positiveOperation, negativeOperation) <- Either.fromOption(
+                getPayloadOperations(operations),
+                InvalidPayloadOperations
+              )
+
+              _ <- Either.cond(
+                positiveOperation.amount.negate.toOption === negativeOperation.amount.some,
+                (),
+                InvalidPayloadOperations
+              )
+
+              transactionAmount <- positiveOperation.amount.value.toTransactionAmount
+                .leftMap(_ => InvalidPayloadOperations)
+
+              sourceAddress = negativeOperation.account.address
+              transactionFee <-
+                metadataResult.suggestedFee match {
+                  case None      => TransactionFee.zero.asRight
+                  case Some(amt) => amt.value.toTransactionFee.leftMap(_ => InvalidSuggestedFee)
+                }
+
+              dagTransaction =
+                DAGTransaction(
+                  source = sourceAddress,
+                  destination = positiveOperation.account.address,
+                  amount = transactionAmount,
+                  fee = transactionFee,
+                  parent = metadataResult.metadata.lastReference,
+                  salt = transactionSalt
+                )
+
+              serializedTxn <-
+                KryoSerializer[F]
+                  .serialize(dagTransaction)
+                  .leftMap(e => SerializationError(e.getMessage))
+
+              txHash <-
```

Comment:
This is not going to be the hash of the `DAGTransaction` but a hash of binary representing the transaction data. If you need the hash of a `DAGTransaction` - the one that is being generated across multiple DAG layers - then you need to run the `.hash` function on the `DAGTransaction` object directly.

Reply:
@TheMMaciek thank you, that helps.